Shopify-analytics: setup basic ECR push workflow

djmetzle opened the pull request:
Add a basic workflow to build and push this image to ECR.

This workflow currently replaces the version in GoCD:
![image](https://github.com/iFixit/shopify-analytics/assets/6876047/3e3437d3-d914-4ffb-8e57-e6e4ffca2ddb)
but is predicated on IAM permissions here: https://github.com/iFixit/server-templates/pull/4260

This accomplishes the main outcome (build and push from Actions instead of GoCD) but is not feature complete. In a few followups, we may want to set up [automatic image updates](https://github.com/iFixit/airflow-dags/blob/master/update-image.py) in the Airflow DAGs repo, to get full end-to-end CD. 

But this accomplishes a replacement of the current build pipeline.

qa_req 0

Ref: https://github.com/iFixit/server-templates/issues/3854

djmetzle committed:
Start a CD workflow

First and foremost, we need to build the image.

djmetzle committed:
Set up ECR Login and push

We'll need to update the terraform in server-templates for IAM to allow
pushing to ECR with the OIDC connector from Github Actions.

andyg0808 reviewed (approved):
CR :+1: 
Looks reasonable to me.

mlahargou commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+
+env:
+  ECR_IMAGE_REPO: "884681002735.dkr.ecr.us-east-1.amazonaws.com/shopify-analytics"
+  DEPLOY_IAM_ROLE: "arn:aws:iam::884681002735:role/github_actions/github-actions-role"
```
I don't see new policies for this role to work on this repo:
https://us-east-1.console.aws.amazon.com/iamv2/home?region=us-west-1#/roles/details/github-actions-role?section=permissions

mlahargou replied:
We also probably need to update the trust relationships to include this repo:
https://us-east-1.console.aws.amazon.com/iamv2/home?region=us-west-1#/roles/details/github-actions-role?section=trust_relationships

djmetzle replied:
Both those concerns are addressed in the PR for the IAM terraform: https://github.com/iFixit/server-templates/pull/4260

mlahargou commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+        uses: aws-actions/amazon-ecr-login@v1
+
+      - name: Push to ECR
+        run: docker push ${{ env.ECR_IMAGE_REPO }}:${{ github.sha }}
```
We can get the registry directly from the ecr login step and thus we don't have to hardcode the `ECR_IMAGE_REPO` variable.

I think it would be something like `${{ steps.login-ecr.outputs.registry }}`

djmetzle replied:
If this was fully automated end-to-end, i'd agree. But we have the URI hardcoded at the consumer: https://github.com/iFixit/airflow-dags/blob/f17de130b8b24e55de248c0a9fb3a101c61eddd6/shopify-analytics/au.py#L40-L41

I like keeping the full URIs around to keep them easier to search for. 

I think eventually we'll want to migrate the Airflow tasks to Dagster, where we can get some better end-to-end CI, and that problem can go away.

mlahargou commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+      - name: Build Image
+        run: docker build -t ${{ env.ECR_IMAGE_REPO }}:${{ github.sha }} --label org.label-schema.name="shopify-analytics" .
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v2
+        with:
+          role-to-assume: ${{ env.DEPLOY_IAM_ROLE }}
+          aws-region: us-east-1
```
I think you need to specify special permissions to log into aws. Something like:

```
      permissions:
        id-token: write
        contents: read
```

Ref: https://github.com/aws-actions/configure-aws-credentials#oidc

djmetzle replied:
Good catch! Thats bit me before too: https://github.com/iFixit/ifixit-bi/pull/88

Thanks!

djmetzle replied:
https://github.com/iFixit/shopify-analytics/pull/13/commits/8aeb3f67dfd0662af73954993cddec5342c07cf1

mlahargou reviewed (dismissed):
dev_block 📱 I don't think this will work as is.

djmetzle committed:
Add permissions for OIDC connector

This workflow needs to modify the "token" used for actions, in order to
get OIDC permission to IAM and ECR. Add this block. The workflow will
fail to login without it.

mlahargou reviewed (approved):
CR 📱

djmetzle commented:
Here goes nothing: https://github.com/iFixit/shopify-analytics/actions/runs/6552509248

djmetzle commented:
Success!
